### PR TITLE
ci(sonarcloud): upload artifact

### PR DESCRIPTION
**Pull request by mmorel-35**
## What does this PR do?

Uploads unit-tests.xml and coverage.out in the sonarcloud artifact during the test phase. After the test-examples job the sonarcloud scan is executed with the artifacts uploaded.

## Why is it important?

This helps providing a correct code coverage on sonarcloud

### CODE REVIEW

File: sonar-project.properties [hunk 1]
```diff
@@ -10,12 +10,12 @@ sonar.projectVersion=v0.23.0
...
 sonar.tests=.
 sonar.test.inclusions=**/*_test.go
 sonar.test.exclusions=**/vendor/**
 
-sonar.go.coverage.reportPaths=coverage.out
-sonar.go.golangci-lint.reportPaths=golangci-lint.report.xml
-sonar.go.tests.reportPaths=TEST-unit.xml
+sonar.go.coverage.reportPaths=**/coverage.out
+sonar.go.golangci-lint.reportPaths=**/golangci-lint.report.xml
+sonar.go.tests.reportPaths=**/TEST-unit.xml
```
Comment (mdelapenya) [inline sonar-project.properties:21]: For modules, we are going to need updating this: https://github.com/testcontainers/testcontainers-go/blob/main/commons-test.mk#L19

```diff
- 		--junitfile TEST-$*.xml \
+ 		--junitfile TEST-unit.xml \
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -139,3 +133,25 @@ jobs:
       rootless-docker: false
       run-tests: true
       ryuk-disabled: false
+
+  sonarcloud:
+    if: ${{ github.ref_name == 'main' && github.repository_owner == 'testcontainers' }}
+    needs: test-examples
```
Comment (mdelapenya) [inline .github/workflows/ci.yml:139]: Do we need to wait until all modules&examples are run? I think we could `need` the main job.

In follow-up PRs, where we add support for the modules, we could move this out to the end (or have one sonarcloud job for each job type: core, modules, examples...) 🤔

Comment (mmorel-35) [reply .github/workflows/ci.yml:139]: There is no obligation but I thought that having only one step to upload every report was better than having one per module.

File: sonar-project.properties (hunk 1 above)
Comment (mmorel-35) [reply sonar-project.properties:21]: Is the file not always named TEST-unit.xml ?

Comment (mdelapenya) [reply sonar-project.properties:21]: I think not for modules :(

File: .github/workflows/ci.yml (hunk 2 above)
Comment (mdelapenya) [reply .github/workflows/ci.yml:139]: Yes, indeed. I meant one job per type: all the modules, all the examples, not for each module or example. I would not consider examples for code coverage, as they are just simple code snippets, not real modules exposing API

Comment (mmorel-35) [reply .github/workflows/ci.yml:139]: Hum... that will make the workflow quite more complex as the upload-artifact will need to receive a dedicated name depending on module, modulegen or root .

File: .github/workflows/ci-test-go.yml [hunk 3]
```diff
@@ -107,15 +107,18 @@ jobs:
...
+        #if: ${{ github.ref_name == 'main' && github.repository_owner == 'testcontainers' && inputs.run-tests && !inputs.rootless-docker }}
+        uses: actions/upload-artifact@v3
         with:
-          projectBaseDir: ${{ inputs.project-directory }}
-        env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
+          name: sonarcloud
+          path: |
+            ./.gitignore
```
Comment (mmorel-35) [inline .github/workflows/ci-test-go.yml:118]: Just a trick to keep all folders

File: .github/workflows/ci-test-go.yml [hunk 4]
```diff
@@ -107,15 +107,17 @@ jobs:
...
+            ${{ inputs.project-directory }}/unit-tests.xml
```
Comment (mdelapenya) [inline .github/workflows/ci-test-go.yml:119]: Is the name of the XML file `unit-tests.xml` or `TEST-unit.xml`?

File: .github/workflows/ci.yml (hunk 2 above)
Comment (mdelapenya) [reply .github/workflows/ci.yml:139]: Ok, one last question: what if the build failed in any of the modules? this step won't be executed. i think we should make it mandatory (as a post-action) for the entire workflow

File: commons-test.mk [hunk 5]
```diff
@@ -16,7 +16,7 @@ test-%:
 		--format short-verbose \
 		--rerun-fails=5 \
 		--packages="./..." \
-		--junitfile TEST-$*.xml \
+		--junitfile unit-tests.xml \
```
Comment (mdelapenya) [inline commons-test.mk:19]: I do prefer `TEST-unit.xml` instead.

File: .github/workflows/ci-test-go.yml (hunk 4 above)
Comment (mdelapenya) [reply .github/workflows/ci-test-go.yml:119]: Never mind, I just saw https://github.com/testcontainers/testcontainers-go/pull/1614/files#diff-f18f81daf1f8305a66be78cac5d122d165f39391f0348477f133858018585cbcR19

File: commons-test.mk (hunk 5 above)
Comment (mmorel-35) [inline commons-test.mk:19]:
```suggestion
		--junitfile TEST-unit.xml \
```

File: .github/workflows/ci-test-go.yml (hunk 4 above)
Comment (mmorel-35) [inline .github/workflows/ci-test-go.yml:119]:
```suggestion
            ${{ inputs.project-directory }}/TEST-unit.xml
```

File: sonar-project.properties [hunk 6]
```diff
@@ -10,12 +10,11 @@ sonar.projectVersion=v0.23.0
...
 
 sonar.tests=.
 sonar.test.inclusions=**/*_test.go
 sonar.test.exclusions=**/vendor/**
 
-sonar.go.coverage.reportPaths=coverage.out
-sonar.go.golangci-lint.reportPaths=golangci-lint.report.xml
-sonar.go.tests.reportPaths=TEST-unit.xml
+sonar.go.coverage.reportPaths=**/coverage.out
+sonar.go.tests.reportPaths=**/unit-tests.xml
```
Comment (mmorel-35) [inline sonar-project.properties:20]:
```suggestion
sonar.go.tests.reportPaths=**/TEST-unit.xml
```

**Review by mdelapenya — COMMENTED** [review]
This LGTM, although before approving I'd like to first discuss https://github.com/testcontainers/testcontainers-go/pull/1614/files#r1324001839

Right after that, I think this is ready to go.

Again, probably this is not adding features to the end-users, but adding lots of value to the maintainability of the project 🚀, which makes a lot for me. Thanks!!!

**Comment by mmorel-35** [comment]
By the way, you’ll need to update the pre release script so the version is also updated in sonar properties

**Review by mdelapenya — APPROVED** [review]

**Comment by mdelapenya** [comment]
> By the way, you’ll need to update the pre release script so the version is also updated in sonar properties

Will do it in a follow-up, thanks!

**Comment by mdelapenya** [comment]
@mmorel-35 here you have it: https://github.com/testcontainers/testcontainers-go/pull/1615

**Pull request merged**